README.md
# BuniCorn Dynamic Automated Market Maker
## Introduction
This repository contains the bunicorn smart contracts.

## Package Manager
We use `yarn` as the package manager. You may use `npm` and `npx` instead, but commands in bash scripts may have to be changed accordingly.

## Requirements
- The following assumes the use of `node@>=10`.

- For interactions or contract deployments on public testnets / mainnet, create a `.env` file specifying your private key and infura api key, with the following format:

```
PRIVATE_KEY=0x****************************************************************
INFURA_API_KEY=********************************
ETHERSCAN_API_KEY=********************************
```

# Setup
For interactions or contract deployments on public testnets / mainnet, create a .env file specifying your private key and infura api key, with the following format:
```
INFURA_API_KEY = 'xxxxx'
ETHERSCAN_API_KEY = 'xxxxx'
PRIVATE_KEY = 'xxxxx'
```

## Install Dependencies

`yarn`

## Compile Contracts

`yarn compile`

## Run Tests

`yarn test`

## Run coverage

`./coverage.sh`

CHANGELOG.md
### Dynamic fee model: change fee calculation based on reserve and volume
- Ema volume is implemented at [here](contracts/VolumeTrendRecorder.sol)
- Fee fomula is implemented at [here](contracts/libraries/FeeFomula.sol)
- The architecture with factory, pair (rename to pool) and router is the same with uniswap model.

Also change:
- Remove price oracle feature.
### AMP model: allow user to amplify the reserve balance of a pool
- For each pair, factory contract can create different pools with the same pair of token.   
Because the price range depends on the initial price, we allow users can create different pools with the same amplification factor.   
But there is only 1 pool with amplification factor = 1
- Router:
    - When users swap from A->B->C, users must specify which pools they choose to swap with at router contract.   
Router contract will also check if these pools are created from factory contract.
    - For add/remove liquidity, users must specify which pools they choose to add/remove liquidity.   
To add liquidity to a new pool, users would use function `addLiquidityNewPool`
- Change in fee:
  - We divide token pools into 4 categories. In general, the higher amplification used indicate greater price stability, so we use lower base fees.
    - Similar assets pools(amplificationFactor > 20): baseFee = 4 bps
    - Strongly correlated assets pools (20 >= amplificationFactor > 5): baseFee = 10 bps
    - Correlated assets pools (5 >= amplificationFactor > 2): baseFee = 20 bps
    - Uncorrelated assets pools: (amplificationFactor <= 2): baseFee = 30 bps
  - For network fee: we keep to mechanism as uniswap but have `FeeToSetter` to set fees up to 20% grow of the pool (uniswap uses 1/6)

test/fee.js
const {artifacts} = require('hardhat');
const BN = web3.utils.BN;
const Helper = require('./helper');

const BuniCornFactory = artifacts.require('BuniCornFactory');
const BuniCornRouter02 = artifacts.require('BuniCornRouter02');
const BuniCornPool = artifacts.require('BuniCornPool');
const FeeTo = artifacts.require('FeeTo');
const MockBuniCornDao = artifacts.require('MockBuniCornDao');
const WBNB = artifacts.require('WBNB9');
const TestToken = artifacts.require('TestToken');

let feeToSetter;
let daoOperator;

let wbnb;
let token;

contract('FeeTo', accounts => {
  before('setup', async () => {
    feeToSetter = accounts[1];
    daoOperator = accounts[2];

    wbnb = await WBNB.new();
    token = await TestToken.new('test', 't1', Helper.expandTo18Decimals(100000));
  });

  it('demo feeTo', async () => {
    let factory = await BuniCornFactory.new(feeToSetter);
    await factory.createPool(wbnb.address, token.address, new BN(10000));
    const poolAddress = await factory.getUnamplifiedPool(wbnb.address, token.address);
    const pool = await BuniCornPool.at(poolAddress);

    /// setup dao and feeTo
    let epoch = new BN(1);
    const dao = await MockBuniCornDao.new(new BN(0), new BN(0), epoch, new BN(0));
    const feeTo = await FeeTo.new(dao.address, daoOperator);
    await factory.setFeeConfiguration(feeTo.address, new BN(1000), {from: feeToSetter});
    await feeTo.setAllowedToken(pool.address, true, {from: daoOperator});

    /// setup router
    let router = await BuniCornRouter02.new(factory.address, wbnb.address);

    await token.approve(router.address, Helper.MaxUint256);
    await router.addLiquidityBNB(
      token.address,
      poolAddress,
      Helper.expandTo18Decimals(100),
      new BN(0),
      new BN(0),
      [Helper.zeroBN, Helper.MaxUint256],
      accounts[0],
      Helper.MaxUint256,
      {value: Helper.expandTo18Decimals(10)}
    );

    let txResult = await router.swapExactBNBForTokens(
      new BN(0),
      [poolAddress],
      [wbnb.address, token.address],
      accounts[0],
      Helper.MaxUint256,
      {
        value: Helper.expandTo18Decimals(1)
      }
    );
    console.log(`gas used when swap with _mintFee: ${txResult.receipt.gasUsed}`);
    /// test gascost with non-zero storage cost
    await dao.advanceEpoch();
    await feeTo.finalize(pool.address, new BN(1));
    txResult = await router.swapExactBNBForTokens(
      new BN(0),
      [poolAddress],
      [wbnb.address, token.address],
      accounts[0],
      Helper.MaxUint256,
      {
        value: Helper.expandTo18Decimals(1)
      }
    );
    console.log(`gas used when swap with _mintFee: ${txResult.receipt.gasUsed}`);

    console.log(await feeTo.rewardsPerEpoch(new BN(1), pool.address));
  });
});


test/flashswap.js
const BN = web3.utils.BN;
const Helper = require('./helper');
const {expandTo18Decimals, MaxUint256} = require('./helper');

const BuniCornRouter = artifacts.require('BuniCornRouter02');
const BuniCornPool = artifacts.require('BuniCornPool');
const BuniCornFactory = artifacts.require('BuniCornFactory');
const WBNB = artifacts.require('WBNB9');
const TestToken = artifacts.require('TestToken');
const ExampleFlashSwap = artifacts.require('ExampleFlashSwap');
const UniswapV2Factory = Helper.getTruffleContract('./node_modules/@uniswap/v2-core/build/UniswapV2Factory.json');
const UniswapV2Router = Helper.getTruffleContract('./node_modules/@uniswap/v2-periphery/build/UniswapV2Router02.json');

let wbnb;
let liquidityProvider;
let trader;
let flashSwap;
let buniFactory;
let buniRouter;
let bnbBuniPool;

let uniswapFactory;
let uniswapRouter;

contract('ExampleFlashSwap', accounts => {
  beforeEach('setup', async () => {
    liquidityProvider = accounts[0];
    trader = accounts[0];
    buniFactory = await BuniCornFactory.new(accounts[0]);
    bnbPartner = await TestToken.new('WBNB Partner', 'WBNB-P', Helper.expandTo18Decimals(10000));
    wbnb = await WBNB.new();
    buniRouter = await BuniCornRouter.new(buniFactory.address, wbnb.address);

    uniswapFactory = await UniswapV2Factory.new(accounts[0]);
    uniswapRouter = await UniswapV2Router.new(uniswapFactory.address, wbnb.address);

    flashSwap = await ExampleFlashSwap.new(uniswapRouter.address, buniFactory.address);

    // await bnbPartner.transfer(trader, initTokenAmount);
    await buniFactory.createPool(wbnb.address, bnbPartner.address, new BN(10000));
    const bnbPoolAddress = await buniFactory.getPools(wbnb.address, bnbPartner.address);
    bnbBuniPool = await BuniCornPool.at(bnbPoolAddress[0]);
  });

  it('uniswapV2Call:0', async () => {
    // add liquidity to uniswap at a rate of 1 BNB / 200 X
    const bnbUniswapPartnerAmount = expandTo18Decimals(2000);
    const bnbUniswapAmount = expandTo18Decimals(10);
    await bnbPartner.approve(uniswapRouter.address, bnbUniswapPartnerAmount);
    await uniswapRouter.addLiquidityBNB(
      bnbPartner.address,
      bnbUniswapPartnerAmount,
      0,
      0,
      liquidityProvider,
      MaxUint256,
      {
        value: bnbUniswapAmount,
        from: accounts[0]
      }
    );

    // add liquidity to buniSwap at a rate of 1 BNB / 100 X
    const bnbPartnerBuniAmount = expandTo18Decimals(1000);
    const bnbBuniAmount = expandTo18Decimals(10);
    await bnbPartner.transfer(bnbBuniPool.address, bnbPartnerBuniAmount);
    await wbnb.deposit({value: bnbBuniAmount});
    await wbnb.transfer(bnbBuniPool.address, bnbBuniAmount);
    await bnbBuniPool.mint(liquidityProvider);

    const balanceBefore = await Helper.getBalancePromise(trader);

    // now, execute arbitrage via buniSwapCall:
    // receive 1 BNB from buniSwap, get as minimum X from uniswap, repay buniSwap with minimum X, keep the rest!
    const arbitrageAmount = expandTo18Decimals(1);
    // instead of being 'hard-coded', the above value could be calculated optimally off-chain. this would be
    // better, but it'd be better yet to calculate the amount at runtime, on-chain. unfortunately, this requires a
    // swap-to-price calculation, which is a little tricky, and out of scope for the moment
    const bnbPoolToken0 = await bnbBuniPool.token0();
    const amount0 = bnbPoolToken0 === bnbPartner.address ? new BN(0) : arbitrageAmount;
    const amount1 = bnbPoolToken0 === bnbPartner.address ? arbitrageAmount : new BN(0);
    await bnbBuniPool.swap(amount0, amount1, flashSwap.address, web3.eth.abi.encodeParameters(['uint'], [new BN(1)]), {
      from: trader,
      gasPrice: new BN(0)
    });

    const balanceAfter = await Helper.getBalancePromise(trader);
    Helper.assertGreater(balanceAfter, balanceBefore);
    console.log(`profit = ${balanceAfter.sub(balanceBefore).toString()}`);
  });

  it('uniswapV2Call:1', async () => {
    // add liquidity to uniswap at a rate of 1 BNB / 100 X
    const bnbUniswapPartnerAmount = expandTo18Decimals(1000);
    const bnbUniswapAmount = expandTo18Decimals(10);
    await bnbPartner.approve(uniswapRouter.address, bnbUniswapPartnerAmount);
    await uniswapRouter.addLiquidityBNB(
      bnbPartner.address,
      bnbUniswapPartnerAmount,
      0,
      0,
      liquidityProvider,
      MaxUint256,
      {
        value: bnbUniswapAmount,
        from: accounts[0]
      }
    );

    // add liquidity to buniSwap at a rate of 1 BNB / 200 X
    const bnbPartnerBuniAmount = expandTo18Decimals(2000);
    const bnbBuniAmount = expandTo18Decimals(10);
    await bnbPartner.transfer(bnbBuniPool.address, bnbPartnerBuniAmount);
    await wbnb.deposit({value: bnbBuniAmount});
    await wbnb.transfer(bnbBuniPool.address, bnbBuniAmount);
    await bnbBuniPool.mint(liquidityProvider);

    const balanceBefore = await Helper.getBalancePromise(trader);

    // now, execute arbitrage via buniSwapCall:
    // receive 200 X from buniSwap, get as much BNB from uniswap as we can, repay buniSwap with minimum BNB, keep the rest!
    const arbitrageAmount = expandTo18Decimals(200);
    // instead of being 'hard-coded', the above value could be calculated optimally off-chain. this would be
    // better, but it'd be better yet to calculate the amount at runtime, on-chain. unfortunately, this requires a
    // swap-to-price calculation, which is a little tricky, and out of scope for the moment
    const bnbPoolToken0 = await bnbBuniPool.token0();
    const amount0 = bnbPoolToken0 === bnbPartner.address ? arbitrageAmount : new BN(0);
    const amount1 = bnbPoolToken0 === bnbPartner.address ? new BN(0) : arbitrageAmount;
    await bnbBuniPool.swap(amount0, amount1, flashSwap.address, web3.eth.abi.encodeParameters(['uint'], [new BN(1)]), {
      from: trader,
      gasPrice: new BN(0)
    });

    const balanceAfter = await Helper.getBalancePromise(trader);
    Helper.assertGreater(balanceAfter, balanceBefore);
    console.log(`profit = ${balanceAfter.sub(balanceBefore).toString()}`);
  });
});


.solcover.js
module.exports = {
  providerOptions: {
    default_balance_ether: 100000000000000,
    total_accounts: 20,
    accounts: [
      // 20 accounts with 10^14 BNB each
      // Addresses:
      //   0xc783df8a850f42e7f7e57013759c285caa701eb6
      //   0xead9c93b79ae7c1591b1fb5323bd777e86e150d4
      //   0xe5904695748fe4a84b40b3fc79de2277660bd1d3
      //   0x92561f28ec438ee9831d00d1d59fbdc981b762b2
      //   0x2ffd013aaa7b5a7da93336c2251075202b33fb2b
      //   0x9fc9c2dfba3b6cf204c37a5f690619772b926e39
      //   0xfbc51a9582d031f2ceaad3959256596c5d3a5468
      //   0x84fae3d3cba24a97817b2a18c2421d462dbbce9f
      //   0xfa3bdc8709226da0da13a4d904c8b66f16c3c8ba
      //   0x6c365935ca8710200c7595f0a72eb6023a7706cd
      //   0xd7de703d9bbc4602242d0f3149e5ffcd30eb3adf
      //   0x532792b73c0c6e7565912e7039c59986f7e1dd1f
      //   0xea960515f8b4c237730f028cbacf0a28e7f45de0
      //   0x3d91185a02774c70287f6c74dd26d13dfb58ff16
      //   0x5585738127d12542a8fd6c71c19d2e4cecdab08a
      //   0x0e0b5a3f244686cf9e7811754379b9114d42f78b
      //   0x704cf59b16fd50efd575342b46ce9c5e07076a4a
      //   0x0a057a7172d0466aef80976d7e8c80647dfd35e3
      //   0x68dfc526037e9030c8f813d014919cc89e7d4d74
      //   0x26c43a1d431a4e5ee86cd55ed7ef9edf3641e901
      {
        secretKey: '0xc5e8f61d1ab959b397eecc0a37a6517b8e67a0e7cf1f4bce5591f3ed80199122',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0xd49743deccbccc5dc7baa8e69e5be03298da8688a15dd202e20f15d5e0e9a9fb',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x23c601ae397441f3ef6f1075dcb0031ff17fb079837beadaf3c84d96c6f3e569',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0xee9d129c1997549ee09c0757af5939b2483d80ad649a0eda68e8b0357ad11131',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x87630b2d1de0fbd5044eb6891b3d9d98c34c8d310c852f98550ba774480e47cc',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x275cc4a2bfd4f612625204a20a2280ab53a6da2d14860c47a9f5affe58ad86d4',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x7f307c41137d1ed409f0a7b028f6c7596f12734b1d289b58099b99d60a96efff',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x2a8aede924268f84156a00761de73998dac7bf703408754b776ff3f873bcec60',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x8b24fd94f1ce869d81a34b95351e7f97b2cd88a891d5c00abc33d0ec9501902e',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29085',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29086',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29087',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29088',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29089',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908a',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908b',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908c',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908d',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908e',
        balance: '100000000000000000000000000000000'
      },
      {
        secretKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908f',
        balance: '100000000000000000000000000000000'
      }
    ]
  },
  skipFiles: ['mock/', 'interfaces/', 'examples/'],
  istanbulReporter: ['html', 'json']
};


deployment/1.deploy.js
const BuniCornFactory = artifacts.require('BuniCornFactory');
const BuniCornRouter02 = artifacts.require('BuniCornRouter02');

async function main () {
  const chainId = await web3.eth.getChainId();
  const metadata = require(`./metadata/${chainId}.json`);

  const accounts = await web3.eth.getAccounts();
  // We get the contract to deploy
  const factory = await BuniCornFactory.new(accounts[0]);
  console.log('Factory deployed to:', factory.address);

  const router = await BuniCornRouter02.new(factory.address);
  console.log('Router deployed to:', router.address);

  await factory.setRouter(router.address);
  console.log('Deploy completed!');
}

main()
  .then(() => process.exit(0))
  .catch(error => {
    console.error(error);
    process.exit(1);
  });

// kovan
// Factory 0x82eF42Fe3657643CD6bc742d902D28753d16460E
// RouterV2 0xB21b3743B733158c93C11D6A29BDcBF7817c86AE

// bsc
// Factory
// RouterV2

deployment/2.deployStableCoinsPool.js
const BigNumber = require("bignumber.js");

const BuniCornRouter02 = artifacts.require('BuniCornRouter02');

async function main () {
  const token0Sym = 'DAI';
  const token1Sym = 'BUSD';

  const chainId = await web3.eth.getChainId();
  const metadata = require(`./metadata/${chainId}.json`);
  const tokens = metadata.tokens;

  const token0 = tokens[token0Sym];
  const token1 = tokens[token1Sym];

  const accounts = await web3.eth.getAccounts();
  const deadline = parseInt(new Date().getTime() / 1000, 10) * 20 * 60;

  const router = await BuniCornRouter02.at(metadata.routerV2Address);
  console.log(`Deploying ${token0.id} (${token0Sym}) - ${token1.id} (${token1Sym})...`)
  await router.addLiquidityNewPool(
    token0.id,
    token1.id,
    new BigNumber(100 * 10000).toFixed(),
    new BigNumber(10000 * 10**token0.decimals).toFixed(),
    new BigNumber(10000 * 10**token1.decimals).toFixed(),
    new BigNumber(10000 * 10**token0.decimals).toFixed(),
    new BigNumber(10000 * 10**token1.decimals).toFixed(),
    accounts[0],
    deadline
  )

  console.log(`Pool ${token0Sym}-${token1Sym} Created!`);
}

main()
  .then(() => process.exit(0))
  .catch(error => {
    console.error(error);
    process.exit(1);
  });

// kovan
// USDT-USDC 0x5895532D9A9c69Fb27331C0a2FA99deF24Bb3B06
// USDT-DAI 0x70b4746769Bf13CCAedF7F601f346B9DCb556Ebd
// USDT-BUSD 0x6fa0c336C472077a1Fd0A059858A5CDCa4F8CB28
// USDC-DAI 0xcCFa5b333bCCee8BDcB69e7F4E5a5FE1595Ed5Bb
// USDC-BUSD 0x1ff9d78972B108066E8204eCB0d770b7f2294e72
// DAI-BUSD 0xaDB7ea2f64719D932ecAE37Ccd82c13D8E6B800f

// bsc

hardhat.config.js
require('@nomiclabs/hardhat-truffle5');
require('@nomiclabs/hardhat-ethers');
require('@nomiclabs/hardhat-web3');
require('@nomiclabs/hardhat-etherscan');
require('hardhat-contract-sizer');
require('solidity-coverage');

require('dotenv').config();

task('accounts', 'Prints the list of accounts', async () => {
  const accounts = await web3.eth.getAccounts();

  for (const account of accounts) {
    console.log(account);
  }
});

module.exports = {
  solidity: {
    version: '0.6.12',
    settings: {
      optimizer: {
        enabled: true,
        runs: 200
      }
    }
  },
  defaultNetwork: 'hardhat',
  networks: {
    hardhat: {
      blockGasLimit: 12500000,
      accounts: [
        // 20 accounts with 10^14 ETH each
        // Addresses:
        //   0xc783df8a850f42e7f7e57013759c285caa701eb6
        //   0xead9c93b79ae7c1591b1fb5323bd777e86e150d4
        //   0xe5904695748fe4a84b40b3fc79de2277660bd1d3
        //   0x92561f28ec438ee9831d00d1d59fbdc981b762b2
        //   0x2ffd013aaa7b5a7da93336c2251075202b33fb2b
        //   0x9fc9c2dfba3b6cf204c37a5f690619772b926e39
        //   0xfbc51a9582d031f2ceaad3959256596c5d3a5468
        //   0x84fae3d3cba24a97817b2a18c2421d462dbbce9f
        //   0xfa3bdc8709226da0da13a4d904c8b66f16c3c8ba
        //   0x6c365935ca8710200c7595f0a72eb6023a7706cd
        //   0xd7de703d9bbc4602242d0f3149e5ffcd30eb3adf
        //   0x532792b73c0c6e7565912e7039c59986f7e1dd1f
        //   0xea960515f8b4c237730f028cbacf0a28e7f45de0
        //   0x3d91185a02774c70287f6c74dd26d13dfb58ff16
        //   0x5585738127d12542a8fd6c71c19d2e4cecdab08a
        //   0x0e0b5a3f244686cf9e7811754379b9114d42f78b
        //   0x704cf59b16fd50efd575342b46ce9c5e07076a4a
        //   0x0a057a7172d0466aef80976d7e8c80647dfd35e3
        //   0x68dfc526037e9030c8f813d014919cc89e7d4d74
        //   0x26c43a1d431a4e5ee86cd55ed7ef9edf3641e901
        {
          privateKey: '0xc5e8f61d1ab959b397eecc0a37a6517b8e67a0e7cf1f4bce5591f3ed80199122',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0xd49743deccbccc5dc7baa8e69e5be03298da8688a15dd202e20f15d5e0e9a9fb',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x23c601ae397441f3ef6f1075dcb0031ff17fb079837beadaf3c84d96c6f3e569',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0xee9d129c1997549ee09c0757af5939b2483d80ad649a0eda68e8b0357ad11131',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x87630b2d1de0fbd5044eb6891b3d9d98c34c8d310c852f98550ba774480e47cc',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x275cc4a2bfd4f612625204a20a2280ab53a6da2d14860c47a9f5affe58ad86d4',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x7f307c41137d1ed409f0a7b028f6c7596f12734b1d289b58099b99d60a96efff',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x2a8aede924268f84156a00761de73998dac7bf703408754b776ff3f873bcec60',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x8b24fd94f1ce869d81a34b95351e7f97b2cd88a891d5c00abc33d0ec9501902e',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29085',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29086',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29087',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29088',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b29089',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908a',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908b',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908c',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908d',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908e',
          balance: '100000000000000000000000000000000'
        },
        {
          privateKey: '0x28d1bfbbafe9d1d4f5a11c3c16ab6bf9084de48d99fbac4058bdfa3c80b2908f',
          balance: '100000000000000000000000000000000'
        }
      ]
    },
    tbsc: {
      chainId: 97,
      gasPrice: 20000000000,
    },
    bsc: {
      chainId: 56,
      gasPrice: 20000000000,
    }
  },
  mocha: {
    enableTimeouts: false
  },
  paths: {
    sources: './contracts',
    tests: './test'
  }
};

const INFURA_API_KEY = process.env.INFURA_API_KEY;
const PRIVATE_KEY = process.env.PRIVATE_KEY;

module.exports.networks.tbsc = {
  url: 'https://data-seed-prebsc-1-s1.binance.org:8545',
  accounts: [PRIVATE_KEY],
  timeout: 20000
};

module.exports.networks.bsc = {
  url: 'https://bsc-dataseed.binance.org',
  accounts: [PRIVATE_KEY],
  timeout: 20000
};

if (INFURA_API_KEY != undefined && PRIVATE_KEY != undefined) {
  module.exports.networks.kovan = {
    url: `https://kovan.infura.io/v3/${INFURA_API_KEY}`,
    accounts: [PRIVATE_KEY],
    timeout: 20000
  };

  module.exports.networks.rinkeby = {
    url: `https://rinkeby.infura.io/v3/${INFURA_API_KEY}`,
    accounts: [PRIVATE_KEY],
    timeout: 20000
  };

  module.exports.networks.ropsten = {
    url: `https://ropsten.infura.io/v3/${INFURA_API_KEY}`,
    accounts: [PRIVATE_KEY],
    timeout: 20000
  };

  module.exports.networks.mainnet = {
    url: `https://mainnet.infura.io/v3/${INFURA_API_KEY}`,
    accounts: [PRIVATE_KEY],
    timeout: 20000
  };
}

if (process.env.ETHERSCAN_API_KEY != undefined) {
  module.exports.etherscan = {
    // Your API key for Etherscan
    // Obtain one at https://etherscan.io/
    apiKey: `${process.env.ETHERSCAN_API_KEY}`
  };
}
